core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -382,27 +378,43 @@ private int assignNewColumnId() {
   }
 
   private TableMetadata applyChangesToMapping(TableMetadata metadata) {
-    String mappingJson = metadata.property(TableProperties.DEFAULT_NAME_MAPPING, null);
-    if (mappingJson != null) {
+    String existingMappingJson = metadata.property(TableProperties.DEFAULT_NAME_MAPPING, null);
+    TableMetadata ret = metadata;
+    if (existingMappingJson != null) {
       try {
         // parse and update the mapping
-        NameMapping mapping = NameMappingParser.fromJson(mappingJson);
+        NameMapping mapping = NameMappingParser.fromJson(existingMappingJson);
         NameMapping updated = MappingUtil.update(mapping, updates, adds);
 
         // replace the table property
         Map<String, String> updatedProperties = Maps.newHashMap();
         updatedProperties.putAll(metadata.properties());
         updatedProperties.put(TableProperties.DEFAULT_NAME_MAPPING, NameMappingParser.toJson(updated));
 
-        return metadata.replaceProperties(updatedProperties);
+        ret = metadata.replaceProperties(updatedProperties);
 
       } catch (RuntimeException e) {
         // log the error, but do not fail the update
-        LOG.warn("Failed to update external schema mapping: {}", mappingJson, e);
+        LOG.warn("Failed to update external schema mapping: {}", existingMappingJson, e);
       }
     }
 
-    return metadata;
+    // Transform the metrics if they exist
+    if (base != null && base.properties() != null) {
+      Schema newSchema = metadata.schema();
+      List<String> deletedColumns = deletes.stream()
+          .map(i -> schema.findColumnName(i))
```

Comment:
a nit but I would rather it was something like "name" instead of i but it may be my bias that I assume i is an int.

Reply:
+1

We can even simplify it further, I guess:

```
deletes.stream()
   .map(schema::findColumnName)
   .collect(Collectors.toList());
```

Reply:
I think `i` means a column id when we build `renamedColumns`. Maybe, just using `id` would make sense.

```
updates.keySet().stream()
     .filter(id -> !schema.findColumnName(id).equals(newSchema.findColumnName(id)))
     .collect(Collectors.toMap(schema::findColumnName, newSchema::findColumnName));
```